src/sessions/base-session-handler.ts

```diff
@@ -0,0 +1,109 @@
+import { checkHasTransceiverFunctionality } from '../media-utils';
+import { throwSdkError } from '../utils';
+
+export default abstract class BaseSessionHandler {
+  constructor (protected sdk: PureCloudWebrtcSdk, protected sessionManager: SessionManager) {}
+
+  abstract getSessionType (): SessionTypes;
+
+  protected log (level: LogLevels, message: any, details?: any): void {
+    log.call(this.sdk, level, message, details);
+  }
+
+  abstract shouldHandleSession ({ sessionInfo, sessionType }: { sessionInfo?: ISessionInfo, sessionType?: SessionTypes }): boolean;
+
+  async startSession (sessionStartParams: IStartSessionParams): Promise<any> {
+    throwSdkError.call(this.sdk, SdkErrorTypes.not_supported, `sessionType ${sessionStartParams.sessionType} can only be started using the purecloud api`, { sessionStartParams });
+  }
+
+  handlePropose (pendingSession: IPendingSession) {
+    this.sdk.emit('pendingSession', pendingSession);
+
+    setTimeout(() => {
+      this.sessionManager.removePendingSession(pendingSession.id);
+    }, 1000);
+  }
+
+  acceptPendingSession (session: IPendingSession, params?: IAcceptPendingSessionRequest) {
+    this.sessionManager.webrtcSessions.acceptRtcSession(session.id);
+  }
+
+  async handleSessionInit (session: any) {
+    try {
+      this.log(LogLevels.info, 'onSession', session);
+    } catch (e) {
+      // don't let log errors ruin a session
+    }
+
+    session.id = session.sid;
+    const pendingSession = this.sessionManager.getPendingSession(session.id);
+    if (pendingSession) {
+      session.conversationId = pendingSession.conversationId;
+    }
+    this.sessionManager.removePendingSession(session.id);
+
+    session._statsGatherer = new StatsGatherer(session.pc, {
+      session: session.sid,
+      conference: session.conversationId
+    });
+
+    session._statsGatherer.on('stats', (data) => {
+      data.conversationId = session.conversationId;
+      this.log(LogLevels.info, 'session:stats', data);
+    });
+
+    session._statsGatherer.on('traces', (data) => {
+      data.conversationId = session.conversationId;
+      this.log(LogLevels.warn, 'session:trace', data);
+    });
+
+    session.on('change:active', (session, active) => {
+      if (active) {
+        session._statsGatherer.collectInitialConnectionStats();
+      }
+      this.log(LogLevels.info, 'change:active', { active, conversationId: session.conversationId, sid: session.sid });
+    });
+
+    session.on('terminated', (session, reason) => {
+      this.log(LogLevels.info, 'onSessionTerminated', { conversationId: session.conversationId, reason });
+      if (session._outboundStream) {
+        session._outboundStream.getTracks().forEach((t: MediaStreamTrack) => t.stop());
+      }
+      this.sdk.emit('sessionEnded', session, reason);
+    });
+    this.sdk.emit('sessionStarted', session);
+  }
+
+  async endSession (session: any) {
```

Comment:
Does this need to be async?

Reply:
Since this is a base method, there may be stuff that needs to be done async such as api requests so this is just future proofing.

Reply:
I think this totally makes sense. Good suggestion. I'll add that in.